### ISSUE TITLE
Feature Request/Question: Improve downloads?

### ISSUE THREAD

**Issue by LukasThyWalls** [crackededed/Xtra#57]
Hello.

Right now the downloads works, and that's something any other twitch app can't say, but in my experience is a bit weird how it works. Maybe is the only solution available, but this is why i'm asking.

For example, in my experience: I try to download a 6h+ video to the external storage (MicroSD card). What Xtra does is to store in their private app storage in the external storage (_"storage"/Android/data/com.github.andreyasadchy.xtra/files/.downloads_, it is the only thing it can do, as it doesn't ask for special permissions, that can be improved but seeing what problems can introduce that can be looked later) in a folder. It can't go at full download speed (It took between 30~60 min in a 500 Mbps connection!) because what Xtra is creating is chunks of <=10 seconds (About more than 2160 .ts files) alongside a .m3u8 file playlist with all of them in order. I know that part of the reason because is so slow is that because is also slow deleting the video: it took several minutes to delete all of them, and because Xtra is not showing that is still doing it (when you tap delete the video is gone), i swapped out Xtra and still there are more than 1000 files in the Xtra SD private storage.
With the clips doesn't do that (it downloads the entire clip as mp4, using the "url name" as the filename), but highlights and vods is what is doing (creates a folder with the video number as name, put all <10sec chunks as .ts alongside a .m3u8 playlist). It seems that those m3u8 works if they are opened directly inside VLC (with its file browser) but not when they are "shared" from a file explorer, and also the "shorter" ones (i tested with a highlight 30 sec. long) doesn't play well when a "larger" one (>40 min VODs) seems to work better. I think is maybe how VLC seeks the position in a m3u8 made with a lot of chunks, and in a shorter one doesn't work pretty well.

So, summarizing, i'm talking if all this can be improved a bit or is what the API allows:
* First, improve the deleting process, not deleting from the download list until all video chunks are deleted from the download folder. Also a "deletion" percentage should be useful.
* Improve how is downloading the videos. I don't know if it can be made that those downloaded chunks larger and make the download time shorter, or instead of downloading and storing them separately, stick them together (as i know, the .ts files can be merged together without recoding, although it need to be noted that many SD can have file size limitations).
* Make them work properly with other video players can open the option to share the video and cast it (For example, with VLC). As now they seems to work with Clips and long VODs/Highlights.
* As a far-off improvement, allow to download to other folders in SD and not only the app private storage (_"storage"/Android/data/com.github.andreyasadchy.xtra/files/.downloads_). It should be dealing with SAF (Storage Access Framework) and as i see in other projects i could be a little annoying to deal with it depending of how it actually do the things and how need to be ported.

Thanks in advance.

**Comment by CrimsonBloodfang** [comment]
I agree with improving downloads. The pause and resume button from the notification sometimes doesn't even work. Having a proper download tab with wider download options is needed like I want to download the .ts files with 32 number of parts to maximize the download speed, or at least unlimited retries when downloads get interrupted, etc. Something similar to what 1DM+ does. External SD Cards can only store files with a 4GB filesize and can't exceed more than that which is why I like the file by file downloads but at the same time we can just change the SD Card filesystem to exFAT to bypass these restrictions, but in my experience it has problems coming with it.

**Comment by crackededed** [comment]
i've added an option to set max concurrent downloads. should speed up the downloads hopefully

**Comment by crackededed** [comment]
i think this is all done now. there is an option to download all chunks from a vod to one file. maybe that works better with vlc
